File: orchestrator/agents/docs_agent.py
"""Docs agent: writes real Markdown to disk describing what changed, pulled
straight from the design node's output so the docs cannot drift from the
API contract that was actually implemented.
"""
import os
from datetime import datetime, timezone


def write_change_doc(ctx, node, design_node_id: str, project_root: str, doc_rel_path: str) -> dict:
    design = ctx.get_output(design_node_id)
    lines = [
        f"## {design['feature_name']}",
        "",
        f"_Generated {datetime.now(timezone.utc).isoformat()} by the docs agent, run `{ctx.run_id}`._",
        "",
        "### API contract",
        "",
    ]
    for endpoint in design["api_contract"]:
        lines.append(f"- `{endpoint['method']} {endpoint['path']}` -- {endpoint['description']}")
    lines += ["", "### Data model changes", ""]
    for change in design["data_model_changes"]:
        lines.append(f"- {change}")
    lines += ["", "### Known risks", ""]
    for risk in design["risks"]:
        lines.append(f"- {risk}")
    lines.append("")
    content = "\n".join(lines)

    abs_path = os.path.join(project_root, doc_rel_path)
    os.makedirs(os.path.dirname(abs_path), exist_ok=True)
    mode = "a" if os.path.exists(abs_path) else "w"
    with open(abs_path, mode, encoding="utf-8") as f:
        if mode == "w":
            f.write(f"# API Changelog\n\n{content}")
        else:
            f.write("\n---\n\n" + content)
    ctx.add_artifact(doc_rel_path)

    ctx.record_decision(
        stage=node.stage,
        node_id=node.id,
        summary=f"Documented '{design['feature_name']}' in {doc_rel_path}",
        rationale="Doc content is generated directly from the design node's approved API contract, so it "
        "cannot describe an endpoint that wasn't actually implemented.",
    )
    return {"doc_path": doc_rel_path, "content": content}

File: src/main/java/com/example/urlshortener/orchestrator/agents/DocsAgent.java
package com.example.urlshortener.orchestrator.agents;

import com.example.urlshortener.orchestrator.Node;
import com.example.urlshortener.orchestrator.SharedContext;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Docs agent: writes real Markdown to disk describing what changed, pulled
 * straight from the design node's output so the docs cannot drift from the
 * API contract that was actually implemented. Direct port of
 * orchestrator/agents/docs_agent.py.
 */
public final class DocsAgent {
    private DocsAgent() {}

    @SuppressWarnings("unchecked")
    public static Map<String, Object> writeChangeDoc(
            SharedContext ctx, Node node, String designNodeId, Path projectRoot, String docRelPath
    ) {
        Map<String, Object> design = ctx.getOutput(designNodeId);
        StringBuilder sb = new StringBuilder();
        sb.append("## ").append(design.get("feature_name")).append("\n\n");
        sb.append("_Generated ").append(Instant.now()).append(" by the docs agent, run `").append(ctx.runId).append("`._\n\n");
        sb.append("### API contract\n\n");
        for (Object o : (List<Object>) design.get("api_contract")) {
            DesignAgent.Endpoint e = (DesignAgent.Endpoint) o;
            sb.append("- `").append(e.method()).append(" ").append(e.path()).append("` -- ").append(e.description()).append("\n");
        }
        sb.append("\n### Data model changes\n\n");
        for (Object c : (List<Object>) design.get("data_model_changes")) sb.append("- ").append(c).append("\n");
        sb.append("\n### Known risks\n\n");
        for (Object r : (List<Object>) design.get("risks")) sb.append("- ").append(r).append("\n");
        sb.append("\n");

        String content = sb.toString();
        Path abs = projectRoot.resolve(docRelPath);
        try {
            Files.createDirectories(abs.getParent());
            if (Files.exists(abs)) {
                Files.writeString(abs, "\n---\n\n" + content, java.nio.file.StandardOpenOption.APPEND);
            } else {
                Files.writeString(abs, "# API Changelog\n\n" + content);
            }
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        ctx.addArtifact(docRelPath);

        ctx.recordDecision(
                node.stage, node.id,
                "Documented '" + design.get("feature_name") + "' in " + docRelPath,
                "Doc content is generated directly from the design node's approved API contract, so it cannot "
                        + "describe an endpoint that wasn't actually implemented."
        );

        Map<String, Object> out = new LinkedHashMap<>();
        out.put("doc_path", docRelPath);
        out.put("content", content);
        return out;
    }
}
